### PR TITLE
[eBPF] Remove libbpf dependencies(#622)

### PR DESCRIPTION
**Pull request by yinjiping**
Since the open source licenses of Libbpf project is GPL, and the open source licenses of
our DeepFlow project is Apache 2.0, the two licenses are incompatible. This patch will
remove the dependence on Libbpf, so we make the following adjustments:

- We rebuilt a framework for loading eBPF binaries (mainly including logical processing for
  ELF, object, program, map, BTF vmlinux file).

- The eBPF bytecode loading process relies on the Libbcc_bpf.a, the static library provided
  by bcc (this library provides a low-level interface such as loading or attaching, and bcc's
  open source licenses is Apache 2.0).

- By analyzing the loaded BTF vmlinux file, the offset of structure members is calculated to
  solve the kernel adaptation problem.
  Note: The kernel must support BTF or a vmlinux file with debug information, otherwise the eBPF
  program will automatically use the feature offset to retry adaptation.

- Some interfaces have been rewritten. e.g.: sysfs cpus is used to resolve and obtain the
  linux kernel version.

### This PR is for:

- Agent

#### Affected branches
- main

### CODE REVIEW

File: agent/build_cn.md [hunk 1]
```diff
@@ -40,6 +40,11 @@ ln -s /usr/bin/llvm-strip-11 /usr/bin/llvm-strip
 
 编译依赖的静态库：
 ```bash
+# bcc
+# 详细参考：[https://github.com/iovisor/bcc/blob/master/INSTALL.md](https://github.com/iovisor/bcc/blob/master/INSTALL.md)
```
Comment (sharang) [inline agent/build_cn.md:44]: 避免中文

Comment (sharang) [reply agent/build_cn.md:44]: 反引号包裹的code中避免中文。

File: agent/src/ebpf/kernel/include/bpf_endian.h [hunk 2]
```diff
@@ -1,19 +1,3 @@
-/*
- * Copyright (c) 2022 Yunshan Networks
```
Comment (sharang) [inline agent/src/ebpf/kernel/include/bpf_endian.h:2]: 为何删掉了

File: agent/src/ebpf/kernel/socket_trace.c [hunk 3]
```diff
@@ -205,16 +192,7 @@ struct skc_flags_t {
 		return SOCK_CHECK_TYPE_ERROR;
 	}
 
-#ifdef BPF_USE_CORE
-	extern __u32 LINUX_KERNEL_VERSION __kconfig;
-	if (LINUX_KERNEL_VERSION >= KERNEL_VERSION(5, 6, 0))
-		bpf_core_read(&conn_info->sk_type, sizeof(conn_info->sk_type), &__sk->sk_type);
```
Comment (nrjatyunshan) [inline agent/src/ebpf/kernel/socket_trace.c:211]: 移除CORE后,后续是不是需要根据内核的变更持续增加需要探测的偏移量

File: agent/src/ebpf/user/elf.c [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+	if (elf_version(EV_CURRENT) == EV_NONE) {
+		goto faild;
+	}
+
+	if (!(*elf = elf_begin(*fd, ELF_C_READ, 0))) {
+		goto faild;
+	}
+
+	return 0;
+faild:
```
Comment (nrjatyunshan) [inline agent/src/ebpf/user/elf.c:47]: faild->failed

File: agent/src/ebpf/user/elf.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	struct sec_desc *map_sec;
+	struct sec_desc *btf_map_sec;
+	struct sec_desc *syms_sec;
+	struct sec_desc *btf_sec;
+	struct sec_desc *btf_ext_sec;
+	struct sec_desc *license_sec;
+	struct sec_desc *version_sec;
+};
+
+int openelf(const char *path, Elf ** elf, int *fd);
```
Comment (nrjatyunshan) [inline agent/src/ebpf/user/elf.h:59]: 有没有 closeelf,这里面打开的文件在哪里关掉的

File: agent/src/ebpf/Makefile [hunk 6]
```diff
@@ -88,12 +73,10 @@ $(ELFFILES):
 	@gcc tools/ebpftobuffer.c -o tools/ebpftobuffer
 	$(call compile_socket_trace_elf, common)
 	$(call compile_socket_trace_elf, 5_2, LINUX_VER_5_2=1)
-	$(call compile_socket_trace_elf, core, CORE=1)
```
[the remark this answers is not in the document]
Comment (yinjiping) [reply agent/src/ebpf/Makefile:91]: 是的

File: agent/src/ebpf/kernel/socket_trace.c (hunk 3 above)
Comment (yinjiping) [reply agent/src/ebpf/kernel/socket_trace.c:211]: 我们还是使用CORE（不想以前那样使用）， 这种方式是从BTF vmlinux中获取偏移值（和先前的原理是一样的），只不过这里简略了，不像以前那么复杂

File: agent/src/ebpf/user/elf.h (hunk 5 above)
Comment (yinjiping) [reply agent/src/ebpf/user/elf.h:59]: 只有fetch_go_elf_version() 中使用了 openelf(), 这个函数实际上是 fd ,elf获取的封装。
fetch_go_elf_version() 函数内部自行释放：
```c
exit:
        if (e)
                elf_end(e);
        if (fd >= 0)
                close(fd);
```

**Review by sharang — DISMISSED** [review]

File: agent/src/ebpf/test/test_elf.c [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        printf("--- struct tcp_sock->write_seq %x\n", offset);
+
+        offset = kernel_struct_field_offset(obj, "task_struct", "files");
+        printf("--- struct task_struct->files %x\n", offset);
+
+        offset = kernel_struct_field_offset(obj, "sock", "__sk_flags_offset");
+        printf("--- struct sock->__sk_flags_offset %x\n", offset);
+
+	for(;;)
+		sleep(1);
```
Comment (nrjatyunshan) [inline agent/src/ebpf/test/test_elf.c:67]: 这里是不是需要去掉

**Review by nrjatyunshan — APPROVED** [review]

**Pull request merged**